R/diffResult.R

```diff
@@ -0,0 +1,194 @@
+            
+            if(self$result == "Passed"){
+                line <- "All checks have passed"
+            } else {
+                line <- "Not all checks have passed, details are provided below"
+            }
+            
+            self$summary <- display$new(
+                title = sprintf("Comparison of Base vs Compare"),
+                body = list(
+                    "Dataset Summary",
+                    dfsum,
+                    "",
+                    "Listing of Keys",
+                    keysum,
+                    "",
+                    "Check Summary",
+                    results,
+                    "",
+                    line,
+                    ""
+                )
+            )
+        }
+    )
+)
+
+
+
+
+#' print.diffResult
+#' 
+#' Print method for a diffdf result object (diffResult). Can be rendered
+#' into different output formats and saved into a report file.
+#' 
+#' @param x diffResult Object
+#' @param type Output format. Options = "ascii" (default) or "html"
+#' @param rowlimit How many rows of a display dataset should be shown (default = 10)
+#' @param file Location and name of a text file to output the results to. Setting to NULL will cause no file to be produced.
+#' @param display Whether to print the rendered output in the R console
+#' @param ... Additional arguments (not used)
+#' 
+#' @export
+print.diffResult <- function(
+    x, 
+    type = "ascii", 
+    rowlimit = 10, 
+    file = NULL, 
+    display = is.null(file), 
+    ...
+){
+        
+    stopifnot(
+        any(type %in% c("ascii", "html"))
+    )
+    
+    if( type == "ascii") render <- render_ascii
+    if( type == "html") render <- render_html
+    
+    failed_displays <- Filter(
+        function(x){ x$result == "Failed" },
+        x$checks
+    )
+    
+    failed_displays <- Map( 
+        function(x) x$display,
+        failed_displays
+    )
+    
+    displays <- append(x$summary, failed_displays)
+    
+    strings_list <- lapply(
+        displays,
+        function(x) x$as.character(render, rowlimit = rowlimit)
+    )
+    
+    strings = unlist(strings_list, use.names = FALSE)
+    
+    if(display) render$print(strings)
+    if (!is.null(file)) render$file(file, strings)
+    
+    return(invisible(strings))
+}
+
+
+
+
+#' @export
+#' @rdname print.diffResult
+as.character.diffResult <- function(x, ...){
+    x <- print(x, display = FALSE, ...)
+    return(x)
+}
+
+
+
+
+#' summary.diffResult
+#' 
+#' Creates a summary of the issues found during the comparison
+#' Returns a list of the issues in a format more accessible
+#' for querying and exploring interactively
+#' 
+#' @param object A diffResult object created by diffdf()
+#' @param ... Additional arguments (not used)
+#' 
+#' @export
+summary.diffResult <- function(object, ...){
+    x <- list()
+    for( i in object$checks){
+        if( i[["result"]] == "Failed"){
+            x[[i[["name"]]]] <- i$data
```

Comment:
I want to consider still returning tibbles. I am reluctant to return data tables because they have behaviour a user will not expect. I think tibbles are nicer objects for non expert users to handle

Reply:
I really wanted to avoid returning tibbles so that we can remove tibble as a dependency.  2/3 of our backwards compatibility issues were due to tibble. I'd rather return data.frames if we could get away with it but we do need support for list columns

Reply:
After sleeping on this and thinking about it some more I think you are right :(  Unfortunately data.table is a niche package in comparison to tibbles and the tidyverse so is likely to confuse or alienate potential users.  I still believe the diffdf internals should strictly use data.table though so I have implemented this as part of the `summary()` function. In general we should really push users to be using that as their primary interface now with the result for inspection.

I have made the relevant changes in a436f0b

Additionally I also updated the print method to cat out the results instead of just printing the table and also added a bunch of unit tests for the summary function.